### PR TITLE
Make decidable versions of sublist functions the default

### LINKED ISSUE #2172: Change in definition of `Data.List.Base.filter` etc. result in new unsolved metas in proofs
The definition of `filter` etc got changed to be defined in terms of the boolean test rather than a decidable predicate. Several problems have resulted:

1. Firstly this is not documented in the CHANGELOG
2. This has lead to many of the proofs not being able to infer their list argument. In particular `filter-accept` and `filter-reject` in `Data.List.Properties` now need their `xs` parameter to be passed explicitly. 

We should check all functions which have been redefined similarly.

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
Fixes #2172 but as discussed in issue, unfortunately introduces an undesirable new dependency from a `Base` file on a `Properties` file.

Maybe the answer is to move `T?` to `Data.Nullary.Decidable.Core`?

### CODE REVIEW

**Comment by jamesmckinna** [comment]
And along with `T?`, all its (slightly rag-tag collection) properties... so double :+1: to the move!
(but if you do, then please also add the [companion lemma](https://github.com/jamesmckinna/agda-stdlib/blob/f938d1bb6b7e578f66ab9688a2fc564cf6b8f90d/src/Relation/Nullary/Reflects.agda#L39-L41)
```agda
T⁺ : ∀ b → Reflects (T b) b
T⁺ true  = of _
T⁺ false = of id
```
to `Relation.Nullary.Reflects`...)

**Comment by jamesmckinna** [comment]
Oh, and if you do go down the road if updating `Reflects` and `Decidable.Core`, then I'd push (back, again) for #2149 to be merged into v2.0 ...

**Comment by JacquesCarette** [comment]
:+1: on moving to `Data.Nullary.Decidable.Core` (though its property should not be in that file too - that should presumably be in `Data.Nullary.Decidable.Properties`).

In general, I think it is a good idea to move definitions of *predicates* out of `*.Properties` files into `*.Predicates` (or the like) files.

**Comment by jamesmckinna** [comment]
Well, I think the property can now be removed entirely, so no need for a distinct `Data.Nullary.Decidable.Properties` in this instance... the only consideration might be how to deprecate `T?-diag` ... but it's used nowhere in the library, so...?

**Comment by MatthewDaggitt** [comment]
> In general, I think it is a good idea to move definitions of predicates out of *.Properties files into *.Predicates (or the like) files.

I'm not aware of any predicate defined in `Properties` files. Only proofs that they are decidable.

**Comment by jamesmckinna** [comment]
Looping back to last weeks discussion: have you experimented with replacing calls to `does (P?)` with `isYes P?`? There's a comment in `Decidable.Core` to the effect that the use of `isYes` helps the type checker infer `P` from a proof of `isYes P?`? If this is *not* the case, we should clean up the comment, and if so, it might be a simpler solution to your issue?

**Comment by MatthewDaggitt** [comment]
> Looping back to last weeks discussion: have you experimented with replacing calls to does (P?) with isYes P?? There's a comment in Decidable.Core to the effect that the use of isYes helps the type checker infer P from a proof of isYes P?? If this is not the case, we should clean up the comment, and if so, it might be a simpler solution to your issue?

I haven't. It would definitely be a worthwhile experiment, although regardless of whether it works or not, I have to confess I'm still happier with this solution though?

**Comment by jamesmckinna** [comment]
Global comment ahead of a full review: now is the moment to deprecate the style of doing `with` on a `Bool`-valued expression, in favour of vanilla `if_then_else_`, surely? #1937 / #2123 or did we decide not to do this? It seems especially glaring here, given that all the functions are being redefined anyway...

**Comment by MatthewDaggitt** [comment]
> Global comment ahead of a full review: now is the moment to deprecate the style of doing with on a Bool-valued expression, in favour of vanilla if_then_else_, surely? https://github.com/agda/agda-stdlib/issues/1937 / https://github.com/agda/agda-stdlib/issues/2123 or did we decide not to do this? It seems especially glaring here, given that all the functions are being redefined anyway...

I did think about that, but I wasn't sure I wanted to risk accidentally introducing any more meta-inference changes...

**Comment by jamesmckinna** [comment]
> > Global comment ahead of a full review: now is the moment to deprecate the style of doing with on a Bool-valued expression, in favour of vanilla if_then_else_, surely? #1937 / #2123 or did we decide not to do this? It seems especially glaring here, given that all the functions are being redefined anyway...
> 
> I did think about that, but I wasn't sure I wanted to risk accidentally introducing any more meta-inference changes...

Fair point! Correspondingly, I can be too impatient/imagine I can reliably see further than is necessarily the case. Small increments might have to be my watchword...

File: src/Data/Vec/Base.agda [hunk 1]
```diff
@@ -17,7 +17,7 @@ open import Data.These.Base as These using (These; this; that; these)
 open import Function.Base using (const; _∘′_; id; _∘_)
 open import Level using (Level)
 open import Relation.Binary.PropositionalEquality.Core using (_≡_; refl; trans; cong)
-open import Relation.Nullary.Decidable.Core using (does)
+open import Relation.Nullary.Decidable.Core using (does; T?)
```
Comment (jamesmckinna) [inline src/Data/Vec/Base.agda:20]: Actually, no need for this import of `T?` here: it's not used anywhere!

File: src/Data/List/Base.agda [hunk 2]
```diff
@@ -346,141 +346,160 @@ removeAt : (xs : List A) → Fin (length xs) → List A
...
   ... | true  = cons acc (go [] cs)
   ... | false = go (c ∷ acc) cs
 
+wordsByᵇ : (A → Bool) → List A → List (List A)
+wordsByᵇ p = wordsBy (T? ∘ p)
+
+derun : ∀ {R : Rel A p} → B.Decidable R → List A → List A
+derun R? [] = []
+derun R? (x ∷ []) = x ∷ []
+derun R? (x ∷ y ∷ xs) with does (R? x y) | derun R? (y ∷ xs)
```
Comment (jamesmckinna) [inline src/Data/List/Base.agda:461]: Recommend (`style-guide`?) using an `@`-pattern here:
```agda
derun R? (x ∷ xs@(y ∷ _)) with does (R? x y) | derun R? xs
```

File: src/Data/List/Base.agda [hunk 3]
```diff
@@ -346,141 +346,160 @@ removeAt : (xs : List A) → Fin (length xs) → List A
...
 partitionᵇ : (A → Bool) → List A → List A × List A
-partitionᵇ p []       = ([] , [])
-partitionᵇ p (x ∷ xs) = (if p x then Prod.map₁ else Prod.map₂′) (x ∷_) (partitionᵇ p xs)
+partitionᵇ p = partition (T? ∘ p)
+
+span : ∀ {P : Pred A p} → Decidable P → List A → (List A × List A)
+span P? []       = ([] , [])
+span P? (x ∷ xs) with does (P? x)
+... | true  = Prod.map (x ∷_) id (span P? xs)
+... | false = ([] , x ∷ xs)
```
Comment (jamesmckinna) [inline src/Data/List/Base.agda:410]: Margin call, but I'd be tempted to use an `@`-pattern here, for the sake of the `false` branch.

**Review by jamesmckinna — APPROVED** [review]
Two global remarks:
* I'm no longer clear about the convoluted history here, but does any of this need a `CHANGELOG` entry?
* this PR seems only/largely to target *definitions*, rather than also their *properties*, and that's how/why you can get away with not investigating the `if_then_else_` forms rather than the `with` forms you have here. But what about the properties of the (new; former?) definitions? I guess my instincts would have been to investigate, but as with your earlier comment, such investigation can come later, I suppose, *but I think that it should*.

File: src/Data/Vec/Base.agda (hunk 1 above)
Comment (MatthewDaggitt) [reply src/Data/Vec/Base.agda:20]: Used in `countᵇ p = count (T? ∘ p)`

File: src/Data/List/Base.agda (hunk 2 above)
Comment (MatthewDaggitt) [reply src/Data/List/Base.agda:461]: Done

File: src/Data/List/Base.agda (hunk 3 above)
Comment (MatthewDaggitt) [reply src/Data/List/Base.agda:410]: Done

**Comment by MatthewDaggitt** [comment]
> I'm no longer clear about the convoluted history here, but does any of this need a CHANGELOG entry?

So the old change should have had an entry. But this refactoring now means that the implementation is actually the same as it was before, so no entry needed!

> this PR seems only/largely to target definitions, rather than also their properties, and that's how/why you can get away with not investigating the if_then_else_ forms rather than the with forms you have here. But what about the properties of the (new; former?) definitions? I guess my instincts would have been to investigate, but as with your earlier comment, such investigation can come later, I suppose, but I think that it should.

The new definitions (the boolean versions) don't have any properties associated with them, and they were introduced in v2.0 so we don't need to worry about altering their behaviour :+1:

File: src/Data/Vec/Base.agda (hunk 1 above)
Comment (jamesmckinna) [reply src/Data/Vec/Base.agda:20]: 🤦

**Pull request merged**